add bml to test operational status

kantera800 opened the pull request:
This PR introduces new API to retrieve radio operational status from a device in real time.
Issue:
#1158

format:
beerocks_cli bml_get_device_operational_radios <al_mac>

sequence:
- register bml listener to topology events
- send topology query
- wait for response (blocking) using beerocks::promise (RESPONSE_TIMEOUT 5sec) 
- unregister bml listener from topology events
- copy topology response data from shared memory.
- parse returned data and print message to console in the same format as used in prplmesh_utils.sh status.

 Next phase would be to use this API in prplmesh_utils.sh status instead of parsing the logs.

tomereli commented on controller/src/beerocks/master/tasks/bml_task.h:
```diff
@@ -67,6 +67,12 @@ class bml_task : public task {
     };
 
+    struct operation_status_update_event {
+        uint8_t backhaul_connection_status;
```
If you get the event - than why do you need backhaul connection status? Its obviously working...

kantera800 replied:
good point

tomereli commented on controller/src/beerocks/master/tasks/bml_task.h:
```diff
@@ -83,6 +89,7 @@ class bml_task : public task {
         ACS_START_EVENT_AVAILABLE,
         CSA_NOTIFICATION_EVENT_AVAILABLE,
         CAC_STATUS_CHANGED_NOTIFICATION_EVENT_AVAILABLE,
+        OPERATIONAL_STATUS_UPDATE,
```
IMO we should call TOPOLOGY_RESPONSE, the BML triggers topology query and waits for topology response, the name operational is just for the command.

kantera800 replied:
done

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml:
```diff
@@ -20,6 +20,16 @@ cACTION_BML_PING_RESPONSE:
+  buffer_size:
+    _type: uint32_t
+    _length_var: True
+  buffer: 
```
Why buffer and not actual response parameters?

kantera800 replied:
done

Commit:
bcl: add DEV_MAX_RADIOS const

preparative commit
Adding deifinetion of maximum (3)  available radios on a single device
will be used later by sDeviceData.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
tlvf: add sDeviceData structure

preparative commit
Device date should contain device al_mac and a list of its radios status.
using const type for radios parameter since
LENGTH_TYPE_DYNAMIC is not supported yet in tlvf.
opened github issue
 https://github.com/prplfoundation/prplMesh/issues/1266

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
tlvf: add BML_TOPOLOGY messages

preparative commit
ACTION_BML_TRIGGER_TOPOLOGY_QUERY already exists.
Adding messages for topology response and
bml listener’s register/unregister events.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml: add BML_DEVICE_DATA structure

preparative commit.
will be used later for get_operational_radios bml api

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml: fix cACTION_BML_TRIGGER_TOPOLOGY_QUERY

Trying to retrieve parent mac from al_mac (bridge) and use
it as destination is wrong and it doesn’t work.
Should just use al_mac as destination.

Rename the bml methods to more understandable name.

Add protection in case bridge is not yet connected (bus not active),
don’t send query message but send empty bml response instead.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml: implement un/register to topology update messages

Preparative commit.
Topology response message is sent every period of time from the agent.
We don’t want to notify all bml listeners of the events all the time.

Adding the ability to register specific bml listener to topology update
events. Once event received we can unregister.
Adding also bml methods.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml: add device_oper_radios_query method

Used to retrieve device radio information
for “bml_get_device_operational_radios” bml API.

Basically what it does is:
1.	register to topology events
2.	send topology query
3.	wait for response using beerocks::promise (blocking)
4.	unregister to topology events
5.	copy topology response data from shared memory.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
beerocks_cli: add bml_get_device_operational_radios api

Add new API to retrieve radio operational status from a device
in real time.
Trigger bml_get_device_operational_radios() that was added in
previous commit.
Parse returned data and print message to console in the same
format that is used in prplmesh_utils.sh status.
Possible next phase would be to use this API in
prplmesh_utils.sh status instead of parsing the logs.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
db: add topology_updates bml listener type

preparative commit
Add topology_updates to sBmlListener structure and update
 is_bml_listener_exist method.
Will be used later on to determine if bml listener is registered to a
topology_updates type events.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
db: add method to enable topology_updates type

preparative commit

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml_task: handle un/register to topology events

Register event enables topology_updates type of events for a specific bml listener.
Unregister does the opposite.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
son_master: notify bml listeners from 1905_topology_response

Notify bml_task each time 1905 topology response CMDU arrives.
In case there is a bml listener that actually registered to
topology events cACTION_BML_TOPOLOGY_RESPONSE will be sent
In other cases event will be ignored by the bml_task so
in operational mode this will have little affect on cpu usage.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml: handle ACTION_BML_TOPOLOGY_RESPONSE message

Before handling,
Check if there are any awaiting threads for this message.
Also, check if the al_mac from response message is the one we are waiting for.
Save response data and signal back to thread.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml_task: handle TOPOLOGY_RESPONSE_UPDATE event

Parse event data and send ACTION_BML_TOPOLOGY_RESPONSE
 message.
Ignore event if there are no bml listeners that are registered
 to topology events.

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
Use send_bml_event_to_listeners in network_map

There is already a dedicated function why not use it .

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
bml: remove dead code

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

Commit:
db: change error log message to warning

Signed-off-by: Alex Kanter <alexx.kanter@intel.com>

morantr commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1852,6 +1852,45 @@ int bml_internal::register_nw_map_update_cb(BML_NW_MAP_QUERY_CB pCB)
+        message_com::create_vs_message<beerocks_message::cACTION_BML_REGISTER_TOPOLOGY_QUERY>(
+            cmdu_tx);
+
+    if (request == nullptr) {
```
```cpp
if (!request) {
    // code
}
```
Please apply it on the other function as well.

morantr commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1852,6 +1852,53 @@ int bml_internal::register_nw_map_update_cb(BML_NW_MAP_QUERY_CB pCB)
+
+    beerocks_message::sDeviceData DeviceData;
+    m_device_data         = &DeviceData;
+    m_device_data->al_mac = network_utils::mac_from_string(device_data->al_mac);
```
need to add nullptr check before accessing input argument `*device_data`.

morantr commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1852,6 +1852,53 @@ int bml_internal::register_nw_map_update_cb(BML_NW_MAP_QUERY_CB pCB)
+        iRet = -BML_RET_TIMEOUT;
+    }
+
+    unregister_topology_discovery_query();
```
Why did you implement a register/unregister mechanism for a simple request-response operation?
The event registration mechanism is relevant if you wish to receive events without any specific request. I didn't see that you added option to register to such event on the bml_cli.
So, the previous commit should be dropped IMO and all other code that related to this event registration.

tomereli replied:
I agree with Moran here. If its possible to avoid extra mechanism and just be notified on each topology response (doesn't really have to be sent from the BML, can be eny response)

kantera800 replied:
I'm not familiar of any other way to signal bml_cli from son_master_thread (internal controller message)

Also couple of points for your suggestion, 
 1. It sounds wrong to go through the bus to send an internal controller message.
2. topology response sent periodically, do we want to put more load on the CPU  especially when this API will not be used in operational state.
4. currently we only use events to signal bml_task from son_master_thread (ACTION_BML directly not used)
5. we don't have eAction for controller->controller VS message (not much of a problem to add one but there is a reason to it)

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -905,6 +909,14 @@ int cli_bml::bml_connection_map_caller(int numOfArgs)
+    if (numOfArgs != 1)
+        return -1;
+    else
+        return get_device_operational_radios(args.stringArgs[0]);
```
```cpp
if (numOfArgs != 1) {
   return -1;
}
return get_device_operational_radios(args.stringArgs[0]);
```

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1351,6 +1363,28 @@ int cli_bml::connection_map()
+    // Main agent
+    std::cout << ((ret == BML_RET_OK) ? "OK" : "FAIL") << " Main radio agent operational"
+              << std::endl;
+    for (auto radio : device_data.radios) {
```
Please use const refernece.
```for (const auto &radio : device_data.radios)```

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2569,7 +2569,8 @@ bool db::is_bml_listener_exist()
     for (auto it = bml_listeners_sockets.begin(); it < bml_listeners_sockets.end(); it++) {
-        listener_exist = (*it).map_updates || (*it).stats_updates || (*it).events_updates;
+        listener_exist = (*it).map_updates || (*it).stats_updates || (*it).events_updates ||
+                         (*it).topology_updates;
```
While you change it, can you please break the ugliness sequence and change it to like this:
```cpp
bool db::is_bml_listener_exist()
{
    for (const auto &soc : bml_listeners_sockets) {
        bool listener_exist =
            soc.map_updates || soc.stats_updates || soc.events_updates || soc.topology_updates;
        if (listener_exist) {
            return true;
        }
    }
    return false;
}
```

Although, as I have said it some previous comment, the registration for topology event if you only use it as request-response style.

morantr commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2494,6 +2494,31 @@ bool db::get_bml_nw_map_update_enable(Socket *sd)
+        }
+    }
+    return false;
+}
```
Please always use this style:
```cpp
if (error) {
    // handle error
}
// rest of the code
```

Also, if you choose to loop with iterators, please use `it->something` and not `(*it).something`.
But I would prefer that you loop with ranged loops:
```cpp
for (auto &soc : bml_listeners_sockets) {
    // Your code here
}
```

morantr commented on controller/src/beerocks/master/tasks/bml_task.cpp:
```diff
@@ -193,6 +193,32 @@ void bml_task::handle_event(int event_type, void *obj)
+            }
+            state = LISTENING;
+        }
+        break;
```
```cpp
if (!obj) {
    // Handle error
}

// Rest of the code

morantr commented on controller/src/beerocks/master/tasks/bml_task.cpp:
```diff
@@ -297,6 +297,57 @@ void bml_task::handle_event(int event_type, void *obj)
         break;
     }
+    case TOPOLOGY_RESPONSE_UPDATE: {
+        if (obj) {
```
```cpp
if (!obj) {
    // handle error
}
// rest of the code

morantr commented on controller/src/beerocks/master/tasks/bml_task.cpp:
```diff
@@ -297,6 +297,57 @@ void bml_task::handle_event(int event_type, void *obj)
     }
+    case TOPOLOGY_RESPONSE_UPDATE: {
+        if (obj) {
+            auto event_obj = (topology_response_update_event *)obj;
```
please use `reinterpret_cast`.

morantr commented on controller/src/beerocks/master/tasks/bml_task.cpp:
```diff
@@ -297,6 +297,57 @@ void bml_task::handle_event(int event_type, void *obj)
+                auto response =
+                    message_com::create_vs_message<beerocks_message::cACTION_BML_TOPOLOGY_RESPONSE>(
+                        cmdu_tx);
+                if (response == nullptr) {
```
```cpp
if (!response) {
```

morantr commented on controller/src/beerocks/master/tasks/bml_task.cpp:
```diff
@@ -297,6 +297,57 @@ void bml_task::handle_event(int event_type, void *obj)
+                idx++;
+            }
+
+            if (!topology_response_updates_listeners.empty()) {
```
```cpp
if (topology_response_updates_listeners.empty()) {
    break;
}
// Rest of the code

morantr commented on controller/src/beerocks/master/tasks/bml_task.cpp:
```diff
@@ -297,6 +297,57 @@ void bml_task::handle_event(int event_type, void *obj)
+
+                size_t i = 0;
+                for (auto iface : event_obj->radio_interfaces) {
+                    auto iface_mac     = iface.mac();
```
const reference

morantr reviewed (approved):
Most of it is styling so approving.

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -67,6 +67,7 @@ enum eStructsConsts {
     CHANNEL_SCAN_LIST_LENGTH             = 8,
     HT_MCS_SET_SIZE                      = 16,
     VHT_MCS_SET_SIZE                     = 32,
+    DEV_MAX_RADIOS                       = 3,
```
Why do we need this? This should be part of the platform, so the correct location for this is not hard coded define but in UCI (nitpick, we can fix it later).

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -660,6 +660,23 @@ sDeviceInfo:
+  al_mac: sMacAddr
+  radios:
+    _type: sIfaceInfo
+    _length: [ "beerocks::message::DEV_MAX_RADIOS" ] 
```
You should consider declaring this as tlvf class where variable length list is supported. Then there is no need for hard coding max radios.
So:
```
cDeviceData:
  _type: class
  al_mac: sMacAddr
  num_radios: uint8_t
  radios:
    _type: sIface_info
    _length: [ num_radios ]
    _length_var: True

kantera800 replied:
I considered creating beerocks_message_common_classes.yaml and adding it there.
After receiving the message i would still need to use a struct with const array to copy the message (BML_DEVICE_DATA) so not sure there is added value doing it.
Also i'm using sDeviceData in other places.

Prefer to leave it a const array.

tomereli commented on controller/src/beerocks/master/tasks/bml_task.h:
```diff
@@ -83,6 +83,9 @@ class bml_task : public task {
         ACS_START_EVENT_AVAILABLE,
```
I think this can be squashed to the previous commit, not much benefit from standing on its own.

kantera800 replied:
done

tomereli commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -471,6 +471,25 @@ struct BML_EVENT_CAC_STATUS_CHANGED_NOTIFICATION {
+
+/**
+ * Beerocks radio information
+ */
```
Please add proper Doxygen comments explaining each member in this struct.

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.h:
```diff
@@ -47,6 +47,12 @@ class bml_internal : public beerocks::socket_thread {
     // Query the beerocks master for the network map
     int nw_map_query();
 
+    // Register topology discovery query so bml listener could receive the responce event.
```
```suggestion
    // Register topology discovery query so bml listener could receive the response event.
```

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1852,6 +1852,53 @@ int bml_internal::register_nw_map_update_cb(BML_NW_MAP_QUERY_CB pCB)
+    // Initialize the promise for receiving the response
+    beerocks::promise<bool> prmDeviceDataGet;
+    m_prmDeviceDataGet = &prmDeviceDataGet;
+    int iOpTimeout     = RESPONSE_TIMEOUT; // Default timeout
```
Don't use cample case for simple variables, its against the coding style.
```suggestion
    int timeout = RESPONSE_TIMEOUT; // Default timeout
```
(nitpick) since it is already like that in this file..

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1852,6 +1852,53 @@ int bml_internal::register_nw_map_update_cb(BML_NW_MAP_QUERY_CB pCB)
+    m_device_data         = &DeviceData;
+    m_device_data->al_mac = network_utils::mac_from_string(device_data->al_mac);
+
+    register_topology_discovery_query();
```
The function name is not OK - it should be `register_topology_discovery_response()`, you don't register to a "query".

kantera800 replied:
first i wrote it exactly like you propose but then i aligned to `register_nw_map_query_cb` naming.
I'll change it back.

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1852,6 +1852,53 @@ int bml_internal::register_nw_map_update_cb(BML_NW_MAP_QUERY_CB pCB)
+
+    register_topology_discovery_query();
+    // device radio status can be retrieved from topology discovery response
+    trigger_topology_discovery(device_data->al_mac);
```
And this one should be `trigger_topology_discovery_query(device_data->al_mac)`

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1852,6 +1852,53 @@ int bml_internal::register_nw_map_update_cb(BML_NW_MAP_QUERY_CB pCB)
+
+    unregister_topology_discovery_query();
+
+    // Clear the credentials member
```
This is not credentials member, copy paste bug?

tomereli commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -109,6 +109,19 @@ int bml_nw_map_register_update_cb(BML_CTX ctx, BML_NW_MAP_QUERY_CB cb);
 int bml_nw_map_query(BML_CTX ctx);
 
+/**
+ * Query the beerocks for the latest radio status on a device
```
Don't use "beerocks" keyword. 
Consider changing to 
>Query prplmesh for radio status for a specific agent, blocking.

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2494,6 +2494,31 @@ bool db::get_bml_nw_map_update_enable(Socket *sd)
+bool db::set_bml_topology_update_enable(Socket *sd, bool update_enable)
+{
+    if (sd) {
+        for (auto it = bml_listeners_sockets.begin(); it < bml_listeners_sockets.end(); it++) {
```
please use range based loops
```suggestion
        for (const auto &soc : bml_listeners_sockets) {
```

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2494,6 +2494,31 @@ bool db::get_bml_nw_map_update_enable(Socket *sd)
+bool db::get_bml_topology_update_enable(Socket *sd)
+{
+    if (sd) {
+        for (auto it = bml_listeners_sockets.begin(); it < bml_listeners_sockets.end(); it++) {
```
same here, please use range based loop

tomereli commented on controller/src/beerocks/master/tasks/bml_task.h:
```diff
@@ -67,6 +68,11 @@ class bml_task : public task {
         uint8_t cac_completed;
     };
 
+    struct topology_response_update_event {
```
This commit doesn't help on its own, just add the code when you use it.

tomereli replied:
Please add Doxygen comments for every new struct.

tomereli commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -555,6 +555,33 @@ int bml_internal::process_cmdu_header(std::shared_ptr<beerocks_header> beerocks_
+
+            // Signal any waiting threads
+            if (m_prmDeviceDataGet) {
+                if (m_device_data != nullptr) {
```
```suggestion
                if (m_device_data) {
```

tomereli commented on controller/src/beerocks/master/db/network_map.cpp:
```diff
@@ -371,9 +371,7 @@ void network_map::send_bml_nodes_statistics_message_to_listeners(
-            for (auto it = bml_listeners.begin(); it != bml_listeners.end(); ++it) {
-                message_com::send_cmdu(*it, cmdu_tx);
-            }
+            send_bml_event_to_listeners(cmdu_tx, bml_listeners);
```
Nice!

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -964,7 +964,7 @@ std::set<std::string> db::get_node_siblings(const std::string &mac, int type)
     auto parent = get_node(n->parent_mac);
     if (!parent) {
-        LOG(ERROR) << "parent for node " << mac << " does not exist";
+        LOG(WARNING) << "parent for node " << mac << " does not exist";
```
Please explain in the commit message why you changed to WARNING. I really don't get what's the difference.. this commit can be dropped I recon.

kantera800 replied:
Its printed every time 1905_topology_response arrives (its a lot or error msg in the log)
The reason for this error is that GW's al_mac  fails to retrieve a parent  (because it doesn't have one).
Since we apply this method on every al_mac  on a neighbor list this specific case can be fixed but i didn't want to do it as part of my PR.
IMO this is a legit usage of the method  and definitely not an error (i can add this explanation to the PR if you want )

tomereli reviewed (approved):
I have many comments but most are simple enough to fix, so approving at first shot! Nice job!

tomereli commented:
You forgot the most important thing - update prplmesh_utils.sh to use this new BML to check the status instead of grepping the logs!
